File: Sources/ClaudeRelayClient/Views/AgentSparkleIcon.swift
import SwiftUI
import ClaudeRelayKit

/// Small "sparkles" glyph shown before a coding agent's name in session rows.
/// Tinted with the agent's palette color and shimmering (variable-color
/// animation) only while the agent is `.working`; static otherwise.
///
/// Conforms to `Equatable` so SwiftUI can short-circuit redraws inside
/// `ForEach` session lists.
public struct AgentSparkleIcon: View, Equatable {
    public var agentId: String?
    public var agentState: AgentDetectedState?
    public var size: CGFloat

    public init(
        agentId: String? = nil,
        agentState: AgentDetectedState? = nil,
        size: CGFloat = 11
    ) {
        self.agentId = agentId
        self.agentState = agentState
        self.size = size
    }

    private var isWorking: Bool { agentState == .working }

    public var body: some View {
        Image(systemName: "sparkles")
            .font(.system(size: size))
            .foregroundStyle(AgentColorPalette.color(for: agentId))
            .symbolEffect(.variableColor.iterative, options: .repeating, isActive: isWorking)
            .accessibilityHidden(true)
    }

    public static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.agentId == rhs.agentId
            && lhs.agentState == rhs.agentState
            && lhs.size == rhs.size
    }
}

File: CodeRelayAndroid/feature-workspace/src/main/kotlin/relay/feature/workspace/ui/AgentSparkleIcon.kt
package relay.feature.workspace.ui

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import relay.protocol.AgentDetectedState

/**
 * Small "sparkles" glyph shown before a coding agent's name in session rows.
 * Mirrors the Swift `AgentSparkleIcon`: tinted with the agent's palette color
 * and shimmering (alpha pulse) only while the agent is [AgentDetectedState.WORKING];
 * fully opaque and static otherwise.
 *
 * SF Symbols' variable-color animation has no Compose equivalent, so working
 * state is conveyed with an `infiniteRepeatable` alpha pulse instead.
 */
@Composable
fun AgentSparkleIcon(
    agentId: String?,
    agentState: AgentDetectedState?,
    size: Dp = 14.dp,
) {
    val working = agentState == AgentDetectedState.WORKING
    val alpha = if (working) {
        val transition = rememberInfiniteTransition(label = "agent-sparkle")
        val animated by transition.animateFloat(
            initialValue = 1f,
            targetValue = 0.35f,
            animationSpec = infiniteRepeatable(
                animation = tween(durationMillis = 700),
                repeatMode = RepeatMode.Reverse,
            ),
            label = "agent-sparkle-alpha",
        )
        animated
    } else {
        1f
    }

    Icon(
        imageVector = Icons.Filled.AutoAwesome,
        contentDescription = null,
        tint = agentColor(agentId),
        modifier = Modifier
            .size(size)
            .alpha(alpha),
    )
}
